### PR TITLE
[agents docs] add pipelines.md etc

### PR DESCRIPTION
**Pull request by yiyixuxu**

### CODE REVIEW

File: .ai/skills/model-integration/SKILL.md [hunk 1]
```diff
@@ -56,7 +56,7 @@ docs/source/en/api/
 ### Integration checklist
 
 - [ ] Implement transformer model with `from_pretrained` support
-- [ ] Implement or reuse scheduler
+- [ ] Implement or reuse scheduler (before writing a new scheduler / guider, grep `src/diffusers/schedulers/` and `src/diffusers/guiders/` for a close match and extend it with a kwarg instead)
```
Comment (yiyixuxu) [inline .ai/skills/model-integration/SKILL.md]:
```suggestion
- [ ] Implement or reuse scheduler
```

File: .ai/skills/model-integration/SKILL.md [hunk 2]
```diff
@@ -65,6 +65,10 @@ docs/source/en/api/
 - [ ] Run `make style` and `make quality`
 - [ ] Test parity with reference implementation (see `parity-testing` skill)
 
+### Pre-PR self-review
+
+Before opening the PR, walk the diff against [../../review-rules.md](../../review-rules.md) — concrete patterns we catch repeatedly in review. Fixing them before submission makes the PR mergeable faster.
+
```
Comment (yiyixuxu) [inline .ai/skills/model-integration/SKILL.md]:
```suggestion
```

File: .ai/models.md [hunk 3]
```diff
@@ -11,7 +11,7 @@ Linked from `AGENTS.md`, `skills/model-integration/SKILL.md`, and `review-rules.
 
 ## Common model conventions
 
-- Models use `ModelMixin` with `register_to_config` for config serialization
+When adding a new transformer (or reviewing one), skim `transformer_flux.py`, `transformer_flux2.py`, `transformer_qwenimage.py`, `transformer_wan.py` first to establish the pattern. Most conventions (mixin set, file structure, naming, gradient-checkpointing implementation, `_no_split_modules` / `_supports_cache_class` settings, etc.) are easiest to internalize by comparison rather than from a fixed list.
```
Comment (yiyixuxu) [inline .ai/models.md]: I think it works better this way, we provide the list of canonical references for them and explicitly to ask them to derive common conventions themselves

File: .ai/pipelines.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
+# Pipeline conventions and rules
```
Comment (yiyixuxu) [inline .ai/pipelines.md:1]: added a pipeline doc for agent

File: .ai/AGENTS.md [hunk 5]
```diff
@@ -27,13 +29,7 @@ Strive to write code as simple and explicit as possible.
...
 ### Pipelines & Schedulers
 
-- Pipelines inherit from `DiffusionPipeline`
-- Schedulers use `SchedulerMixin` with `ConfigMixin`
-- Use `@torch.no_grad()` on pipeline `__call__`
-- Support `output_type="latent"` for skipping VAE decode
-- Support `generator` parameter for reproducibility
-- Use `self.progress_bar(timesteps)` for progress tracking
-- Don't subclass an existing pipeline for a variant — DO NOT use an existing pipeline class (e.g., `FluxPipeline`) to override another pipeline (e.g., `FluxImg2ImgPipeline`) which will be a part of the core codebase (`src`)
+- See [pipelines.md](pipelines.md) for pipeline conventions, patterns, and gotchas.
```
Comment (sayakpaul) [inline .ai/AGENTS.md:32]: Good call!

File: .ai/models.md (hunk 3 above)
Comment (sayakpaul) [inline .ai/models.md]:
```suggestion
* Models use `ModelMixin` with `register_to_config` for config serialization. 
* When adding a new transformer (or reviewing one), skim `src/diffusers/models/transformers/transformer_flux.py`, `src/diffusers/models/transformers/transformer_flux2.py`, `src/diffusers/models/transformers/transformer_qwenimage.py`, and `src/diffusers/models/transformers/transformer_wan.py` first to establish the pattern. Most conventions (mixin set, file structure, naming, gradient-checkpointing implementation, `_no_split_modules` settings, etc.) are easiest to internalize by comparison rather than from a fixed list.
```

File: .ai/models.md [hunk 6]
```diff
@@ -55,27 +55,28 @@ class MyModelAttention(nn.Module, AttentionModuleMixin):
...
+### Attention masks
+
+What you pass as `attn_mask=` to `dispatch_attention_fn` determines which backends work:
+
+- **No mask needed → pass `None`, not an all-zero tensor.** A dense 4D additive float mask of all `0.0` does no math but still hard-raises on `flash` / `_flash_3` / `_sage` (see `attention_dispatch.py:2328, 2544, 3266`). Only materialize a mask when it carries information. This is the Flux / Flux2 / Wan pattern: no mask, works on every backend, relies on the model having been trained tolerating consistent padding.
+- **Padding mask → bool `(B, L)` or `(B, 1, 1, L)`.** Stays compatible with the `*_varlen` kernels via `_normalize_attn_mask` (`attention_dispatch.py:639`), which reduces bool masks to `cu_seqlens`. Dense additive-float masks *cannot* be reduced this way and so lose the varlen path. This is the Qwen pattern (`transformer_qwenimage.py:951`).
+- **Structural mask (causal, sliding-window, band-diagonal) → dense `(1, 1, L, L)` is unavoidable.** Row-varying patterns can't be expressed as `(B, L)`. Expect SDPA/Flex-only for these layers; consider Flex's `sliding_window_mask_mod` or FA3's native `window_size=` kwarg if backend flexibility matters.
```
Comment (sayakpaul) [inline .ai/models.md]:
```suggestion
- **Structural mask (causal, sliding-window, band-diagonal) → dense `(1, 1, L, L)` is unavoidable.** Row-varying patterns can't be expressed as `(B, L)`. Expect SDPA/Flex-only for these layers; consider Flex's `sliding_window_mask_mod` or FA3's native `window_size=` kwarg if backend flexibility matters. Consult `src/models/transformers/transformer_kandinsky.py` as a reference.
```

File: .ai/models.md (hunk 3 above)
Comment (sayakpaul) [reply .ai/models.md]: I removed `_supports_cache_class` because the attribute doesn't seem to exist on `ModelMixin`.

File: .ai/models.md [hunk 7]
```diff
@@ -55,27 +55,28 @@ class MyModelAttention(nn.Module, AttentionModuleMixin):
...
-
-6. **Config serialization gaps.** Every `__init__` parameter in a `ModelMixin` subclass must be captured by `register_to_config`. If you add a new param but forget to register it, `from_pretrained` will silently use the default instead of the saved value.
-
-7. **Forgetting to update `_import_structure` and `_lazy_modules`.** The top-level `src/diffusers/__init__.py` has both -- missing either one causes partial import failures.
+4. **Capability flags without matching implementation.** Class attributes like `_supports_cache_class`, `_no_split_modules`, `_supports_gradient_checkpointing`, `_supports_quantization` advertise a contract. Setting one without the corresponding logic (e.g. `_supports_gradient_checkpointing = True` with no `_set_gradient_checkpointing` method, or no `if gradient_checkpointing` branches in `forward`) means training code silently no-ops or errors deep inside `torch.utils.checkpoint`. For `_supports_cache_class` / `_no_split_modules` specifically, wrong values cause silent correctness bugs or OOM. Copy from a similar model, verify the corresponding logic is in place, and for inference-only ports just drop the flag.
```
Comment (sayakpaul) [inline .ai/models.md]: `_supports_quantization` -- we don't have this attribute.

File: .ai/models.md [hunk 8]
```diff
@@ -55,27 +55,28 @@ class MyModelAttention(nn.Module, AttentionModuleMixin):
...
+- **Structural mask (causal, sliding-window, band-diagonal) → dense `(1, 1, L, L)` is unavoidable.** Row-varying patterns can't be expressed as `(B, L)`. Expect SDPA/Flex-only for these layers; consider Flex's `sliding_window_mask_mod` or FA3's native `window_size=` kwarg if backend flexibility matters.
+- **Don't declare `attention_mask` (or `encoder_hidden_states_mask`) in the forward signature if you ignore it.** "For API stability with other transformers" is not a reason; readers assume a declared param is honored, and downstream pipelines will pass padding masks that silently get dropped. Note: some existing models in the repo carry unused mask params for historical reasons — e.g. `QwenDoubleStreamAttnProcessor2_0.__call__` declares `encoder_hidden_states_mask` but never reads it (the joint mask is routed through `attention_mask` instead), and the block-level forward in `transformer_qwenimage.py` declares it but always receives `None`. This is legacy and should not be replicated in new models.
```
Comment (stevhliu) [inline .ai/models.md]:
```suggestion
- **Don't declare `attention_mask` (or `encoder_hidden_states_mask`) in the forward signature if you ignore it.** "For API stability with other transformers" is not a reason; readers assume a declared param is honored, and downstream pipelines will pass padding masks that silently get dropped. Some existing models in the repo carry unused mask params for historical reasons — e.g. `QwenDoubleStreamAttnProcessor2_0.__call__` declares `encoder_hidden_states_mask` but never reads it (the joint mask is routed through `attention_mask` instead), and the block-level forward in `transformer_qwenimage.py` declares it but always receives `None`. This is a legacy behavior and should not be replicated in new models.
```

File: .ai/pipelines.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+       self.vae_scale_factor = (
+           2 ** (len(self.vae.config.block_out_channels) - 1) if getattr(self, "vae", None) else 8
+       )
+       self.sample_rate = int(self.vae.config.sampling_rate) if getattr(self, "vae", None) else 48000
+   ```
+   `@property` is reserved for per-call state — values that depend on something set inside `__call__` (e.g. `do_classifier_free_guidance` reading `self._guidance_scale`).
+
+2. **`@torch.no_grad()` discipline.** Two failure modes:
+    - **Missing on `__call__` entirely** — causes GPU OOM from gradient accumulation during inference. Always decorate `__call__` with `@torch.no_grad()`.
+    - **Redundant inside helpers** that `__call__` already covers. The decorator puts every descendant in no-grad, so an inner `with torch.no_grad():` is noise — and worse, it forecloses callers who want to invoke `pipe.encode_prompt(...)` with grads enabled (training, embedding optimization). Convention across diffusers (flux, qwen, flux2, stable_audio, audioldm2) is decorator-only.
```
Comment (stevhliu) [inline .ai/pipelines.md]:
```suggestion
    - **Redundant inside helpers** that `__call__` already covers. The decorator puts every descendent in no-grad, so an inner `with torch.no_grad():` is noise — and worse, it forecloses callers who want to invoke `pipe.encode_prompt(...)` with grads enabled (training, embedding optimization). Convention across diffusers (flux, qwen, flux2, stable_audio, audioldm2) is decorator-only.
```

File: .ai/pipelines.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+3. **Reinventing logic that already exists in the repo.** Before adding new logic to a pipeline, check whether diffusers already has it. Reuse out of the box if you can; extend with a small kwarg if the change is genuinely small (use judgement on whether to extend an existing class or write a new one).
```
Comment (stevhliu) [inline .ai/pipelines.md]:
```suggestion
3. **Reinventing logic that already exists in the repo.** Check `src/diffusers/guiders/` and `src/diffusers/schedulers/` before adding new logic. Reuse what's already there; extend with a small kwarg for minor variations.
```

File: .ai/review-rules.md [hunk 11]
```diff
@@ -5,8 +5,13 @@ Review-specific rules for Claude. Focus on correctness — style is handled by r
...
 - [modular.md](modular.md) — modular pipeline conventions, patterns, common mistakes
 - [skills/parity-testing/SKILL.md](skills/parity-testing/SKILL.md) — testing rules, comparison utilities
 - [skills/parity-testing/pitfalls.md](skills/parity-testing/pitfalls.md) — known pitfalls (dtype mismatches, config assumptions, etc.)
 
-## Common mistakes (add new rules below this line)
+## Common mistakes
+
+Common mistakes are covered in the common-mistakes / gotcha sections in [AGENTS.md](AGENTS.md), [models.md](models.md), [pipelines.md](pipelines.md), and [modular.md](modular.md). Additionally, watch for below patterns that aren't covered there:
+
+- **Ephemeral context.** Comments, docstrings, and files that only made sense to the current PR's author or reviewer don't help a future reader/user/developer. Examples: `# per reviewer comment on PR #NNNN`, `# as discussed in review`, `# TODO from offline chat`, debug printouts. Same for files: parity harnesses, comparison scripts, anything in `scripts/` with hardcoded developer paths or imports from the reference repo. State the *reason* so the comment stands alone, or drop it.
```
Comment (stevhliu) [inline .ai/review-rules.md:17]: nice!

**Review by stevhliu — COMMENTED** [review]
thanks, very well thought out!
